doc/dev/tma.md

```diff
@@ -171,6 +171,251 @@ and the former is not equivalent to the latter as discussed in [Divisibility of
+if a bit in `i` is `0`, then the result of the xor operation will be `j`'s corresponding bit unchanged.
+
+Considering the lowest bit of `i`, it has value `1` every other row.
+When this bit is xor-ed with `j`,
+the result `i ^ j`'s lowest bit will be the flip of the lowest bit of `j` every other row.
+Visually, this corresponding to swapping each pair of unit matrices every other row.
+
+When we treat $2\times 2$ unit matrices as new unit, we are effectively doing `i >> 1` and `j >> 1`.
+Repeat the step 1 will then use the lowest bit of `(i >> 1) ^ (j >> 1)` to control the swapping of the new units.
+This equivalent to use the second lowest bit of `i ^ j` to control the $2\times 2$ unit matrices.
+
+When we repeat this process until finished, we are using all bits of `i ^ j`,
+different bits control the flip of pairs of different granularity.
+
+$\square$
+
+</details>
+
+Storing data in swizzled memory format in shared memory is helpful because it makes both row-wise and column-wise memory access free of bank conflict.
+
+<details>
+
+**<summary>Why?</summary>**
+
+In the 3-step procedure above, we are only swapping items within each row of unit matrices.
+This procedure never interleave different rows, instead,
+it just reorder the unit matrices within each row itself.
+So each row of unit matrices still occupies the full width of 32 banks.
+Therefore, when a warp is simultenuously accessing a row of unit matrices
+(either by vectorization or by parallelization), it has no bank conflict.
+
+In the 3-step approach above, the first step swaps half of the items in the first column with half of the items with the second column.
+This operation interleave every pair of columns to make them containing each other evenly.
+If we consider every 2 columns as a unit column, then items between different unit columns still does not interleave.
+But when we consider $2\times 2$ unit matrices as new unit and do the swap again,
+we are interleaving every pair of unit columns making them containing each other evenly.
+If we repeat this process, each logical column will eventually be evenly distributed to every physical column.
+So when we want to access one logical column, we are actually accessing every physical columns.
+Therefore, when a warp is simultenuously accessing a logical column of unit matrices,
+this access will be distributed to all physical columns, therefore all banks.
+Therefore, this access has no bank conflict.
+
+$\square$
+
+</details>
+
+When we use the automatic swizzling feature of TMA,
+we must make sure that the allocation domain of our shared memory tensor is consistent with the memory format defined by TMA.
+
+Let `s` be swizzle size (`bytes of swizzle / 16 bytes`).
+The overall structure of the schedule of the allocation domain of the shared memory tensor looks like:
+
+```
+tile IterDomains => view as ([-1, s, 8/s, s, 16B]) => swizzle(1, 3) => arbitrary view
```

Comment:
Isn't `8/s` always supposed to be 2?

Reply:
For 32 byte swizzle, this is always 4. For 64 byte swizzle, this is always 2. For 128 byte swizzle, this is always 1 (so we can just omit this and schedule a `[-1, s, s, 16B]`).

Reply:
OK, I see. When `n = 32`, the xor swizzle steps is only done once since `32 / 16 = 2`, right?

Reply:
Yes

Reply:
Updated the doc mentioning the concrete values for each swizzle mode.